### PR TITLE
[FEATURE] Calculate code coverage

### PR DESCRIPTION
**Pull request by lukaszuznanski**
Calculate code coverage of extension

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 [![GitHub CI status](https://github.com/TYPO3-Documentation/tea/workflows/CI/badge.svg?branch=main)](https://github.com/TYPO3-Documentation/tea/actions)
 [![GitLab CI status](https://gitlab.typo3.org/qa/example-extension/badges/main/pipeline.svg)](https://gitlab.typo3.org/qa/example-extension/-/pipelines)
 [![Latest stable version](https://poser.pugx.org/ttn/tea/v/stable.svg)](https://packagist.org/packages/ttn/tea)
+[![Code Coverage Badge](./badge.svg)](./badge.svg)
```
Comment (oliverklee) [inline README.md:6]: Shouldn't the badge link to the code coverage report?

File: .gitattributes [hunk 2]
```diff
@@ -11,7 +11,9 @@
 /Resources/Private/package.json export-ignore
 /Resources/Private/stylelint.config.js export-ignore
 /Tests/ export-ignore
+/badge.svg export-ignore
 /codeception.yml export-ignore
+/clover.xml export-ignore
```
Comment (oliverklee) [inline .gitattributes:16]: I'd prefer to git-ignore `clover.xml` instead and keep it out of the repo. (Otherwise, we'd have to include the coverage changes in each PR.)

File: .github/workflows/codecoverage.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
+---
+# This GitHub Actions workflow is calculating code coverage of extension and creates badge out of it.
+name: Code coverage
+on: push
+jobs:
+  code-coverage:
+    name: "Code coverage check"
```
Comment (oliverklee) [inline .github/workflows/codecoverage.yml:7]: This is not a check. It **calculates** the coverage.

File: .github/workflows/codecoverage.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    runs-on: ubuntu-18.04
+    if: "!contains(github.event.head_commit.message, '[ci skip]')"
```
Comment (oliverklee) [inline .github/workflows/codecoverage.yml:9]: I propose we keep this out of this PR and move it to a separate PR so we can have it for the other checks as well.

File: Tests/Functional/phpunit.xml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    stopOnIncomplete="false"
+    stopOnSkipped="false"
+    verbose="false"
+>
+    <testsuites>
+        <testsuite name="Functional">
+            <directory>./</directory>
+        </testsuite>
+    </testsuites>
+</phpunit>
```
Comment (oliverklee) [inline Tests/Functional/phpunit.xml:21]: Can't we use the configuration files from the nimut testing framework and provide the paths to the unit/functional tests on the command line?

File: badge.svg [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<svg width="100" height="20" xmlns="http://www.w3.org/2000/svg">
```
Comment (oliverklee) [inline badge.svg:1]: Could we also create the badge on-the-fly instead of having the badge with the number git-versioned?

File: composer.json [hunk 7]
```diff
@@ -89,6 +89,13 @@
 			"@ci:static"
 		],
 		"ci:composer:normalize": "@composer normalize --dry-run",
+		"ci:coverage": [
```
Comment (oliverklee) [inline composer.json:92]: Please also let's have descriptions for the new commands.

File: phive.xml [hunk 8]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <phive xmlns="https://phar.io/phive">
-    <phar name="phar-io/phive" version="^0.14.5" installed="0.14.5" location="./tools/phive" copy="true"/>
-    <phar name="phpcpd" version="^6.0.3" installed="6.0.3" location="./tools/phpcpd" copy="true"/>
+  <phar name="phar-io/phive" version="^0.14.5" installed="0.14.5" location="./tools/phive" copy="true"/>
+  <phar name="phpcpd" version="^6.0.3" installed="6.0.3" location="./tools/phpcpd" copy="true"/>
+  <phar name="phpcov" version="^5.0.0" installed="5.0.0" location="./tools/phpcov" copy="true"/>
```
Comment (oliverklee) [inline phive.xml:5]: Can we use the Composer-installed version instead?

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: phive.xml (hunk 8 above)
Comment (lukaszuznanski) [reply phive.xml]: unfortunately there is some issue with dependencies, we should use phive for this tool

File: composer.json [hunk 9]
```diff
@@ -163,6 +170,9 @@
 	"scripts-descriptions": {
 		"ci": "Runs all dynamic and static code checks.",
 		"ci:composer:normalize": "Checks the composer.json.",
+		"ci:coverage:functional": "Generates code coverage raport out of functional tests.",
```
[the remark this answers is not in the document]
Comment (oliverklee) [reply composer.json:173]: out of -> for

[the remark this answers is not in the document]
Comment (oliverklee) [reply composer.json:173]: Generates -> Generates the

File: .github/workflows/codecoverage.yml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      - name: "Run unit tests coverage"
+        run: composer ci:coverage:unit
+      - name: "Merge coverage results"
+        run: composer ci:coverage:merge
+      - name: phpunit-coverage-badge
+        uses: timkrase/phpunit-coverage-badge@v1.2.0
+        with:
+          coverage_badge_path: './coverage/badge.svg'
+      - name: "Copy result to coverage directory"
+        run: cp clover.xml ./coverage/clover.xml
```
Comment (oliverklee) [inline .github/workflows/codecoverage.yml:54]: Do we need this, or could we directly generate the merged covered `clover.xml` in `coverage/`?

File: README.md [hunk 11]
```diff
@@ -3,6 +3,7 @@
 [![GitHub CI status](https://github.com/TYPO3-Documentation/tea/workflows/CI/badge.svg?branch=main)](https://github.com/TYPO3-Documentation/tea/actions)
 [![GitLab CI status](https://gitlab.typo3.org/qa/example-extension/badges/main/pipeline.svg)](https://gitlab.typo3.org/qa/example-extension/-/pipelines)
 [![Latest stable version](https://poser.pugx.org/ttn/tea/v/stable.svg)](https://packagist.org/packages/ttn/tea)
+[![Code Coverage Badge](https://github.com/TYPO3-Documentation/tea/blob/code-coverage-badge/badge.svg)](https://github.com/TYPO3-Documentation/tea/blob/code-coverage-badge/clover.xml)
```
[the remark this answers is not in the document]
Comment (lukaszuznanski) [reply README.md:6]: but then we need to keep result of code coverage on main branch

File: .github/workflows/codecoverage.yml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+          BRANCH: code-coverage-badge
+          FOLDER: coverage
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    strategy:
+      fail-fast: false
+      matrix:
+        php-version:
+          - 7.4
+        typo3-version:
+          - ^10.4
```
[the remark this answers is not in the document]
Comment (lukaszuznanski) [reply .github/workflows/codecoverage.yml:67]: I think we should make sure that tests are running properly with version 11.5 and then upgrade it here also

File: README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (oliverklee) [reply README.md:6]: Ah, okay. Makes sense.

File: composer.json [hunk 13]
```diff
@@ -163,6 +170,9 @@
 	"scripts-descriptions": {
 		"ci": "Runs all dynamic and static code checks.",
 		"ci:composer:normalize": "Checks the composer.json.",
+		"ci:coverage:functional": "Generates the code coverage report for functional tests.",
+		"ci:coverage:merge": "Merges the code coverage reports form unit and functional tests.",
```
Comment (oliverklee) [inline composer.json:174]:
```suggestion
		"ci:coverage:merge": "Merges the code coverage reports for unit and functional tests.",
```
